baseplate/config.py

```diff
@@ -318,62 +338,73 @@ def one_of(text):
     return one_of
 
 
-def TupleOf(T):  # noqa: D401
+def TupleOf(item_parser: Callable[[str], T]) -> Callable[[str], Sequence[T]]:  # noqa: D401
     """A comma-delimited list of type T.
 
     At least one value must be provided. If you want an empty list
     to be a valid choice, wrap with :py:func:`Optional`.
 
     """
 
-    def tuple_of(text):
+    def tuple_of(text: str) -> Sequence[T]:
         if not text:
             raise ValueError("no values provided")
         split = text.split(",")
         stripped = [item.strip() for item in split]
-        return [T(item) for item in stripped if item]
+        return [item_parser(item) for item in stripped if item]
 
     return tuple_of
 
 
-def Optional(T, default=None):  # noqa: D401
+def Optional(
+    item_parser: Callable[[str], T], default: OptionalType[T] = None
+) -> Callable[[str], OptionalType[T]]:  # noqa: D401
     """An option of type T, or ``default`` if not configured."""
 
-    def optional(text):
+    def optional(text: str) -> OptionalType[T]:
         if text:
-            return T(text)
+            return item_parser(text)
         return default
 
     return optional
 
 
-def Fallback(T1, T2):  # noqa: D401
+def Fallback(
+    primary_parser: Callable[[str], T], fallback_parser: Callable[[str], T]
+) -> Callable[[str], T]:  # noqa: D401
     """An option of type T1, or if that fails to parse, of type T2.
 
     This is useful for backwards-compatible configuration changes.
 
     """
 
-    def fallback(text):
+    def fallback(text: str) -> T:
         try:
-            return T1(text)
+            return primary_parser(text)
         except ValueError:
-            return T2(text)
+            return fallback_parser(text)
 
     return fallback
 
 
 class ConfigNamespace(dict):
     def __init__(self):
-        super(ConfigNamespace, self).__init__()
+        super().__init__()
         self.__dict__ = self
 
+    def __getattr__(self, name: str) -> Any:
+        ...
```

Comment:
What's this?

Reply:
It's a type stub for `__getattr__` so that type checking knows we can return anything we want via attribute references. Prior art is here:

* https://github.com/python/typeshed/blob/master/stdlib/2and3/argparse.pyi#L268
* https://github.com/python/typeshed/issues/25

Reply:
tl;dr "good luck!!!!!! i'll do whatever i want!!!!"